thanos: init at 0.6.0 & NixOS module

basvandijk opened the pull request:
~**Please don't merge since this is work-in-progress.**~ EDIT: this is now ready for review.

<!-- Nixpkgs has a lot of new incoming Pull Requests, but not enough people to review this constant stream. Even if you aren't a committer, we would appreciate reviews of other PRs, especially simple ones like package updates. Just testing the relevant package/service and leaving a comment saying what you tested, how you tested it and whether it worked would be great. List of open PRs: <https://github.com/NixOS/nixpkgs/pulls>, for more about reviewing contributions: <https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#sec-reviewing-contributions>. Reviewing isn't mandatory, but it would help out a lot and reduce the average time-to-merge for all of us. Thanks a lot if you do! -->
###### Motivation for this change

I would like to use [Thanos](https://github.com/improbable-eng/thanos/blob/master/README.md) in NixOS.

[0.4.0 Release notes](https://github.com/improbable-eng/thanos/releases/tag/v0.4.0).

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

basvandijk committed:
buildGoModule: support overriding the go-modules derivation

basvandijk committed:
thanos: init at 0.4.0

basvandijk committed:
nixos/thanos: add module for the thanos service

basvandijk commented:
@GrahamcOfBorg build thanos

The following contains some preliminary tests for thanos:
@GrahamcOfBorg test prometheus2

basvandijk commented:
@Mic92 I see you've done some work on `buildGoModule`. It would be great if you could review commit https://github.com/NixOS/nixpkgs/pull/60500/commits/c93e5ca96a53d79b6c72552105302f271cc19c92. The `thanos` go package needs to have access to the `bzr` executable when downloading the go module dependencies. This commit adds a way to override the `go-modules` derivation so that we can [add `bzr` to the `nativeBuildInputs`](https://github.com/NixOS/nixpkgs/blob/ff7e7e957d79318b17bbb2c5bf5b4df3181a9a7b/pkgs/servers/monitoring/thanos/default.nix#L20:L24).

Infinisil commented:
I really don't like having more such huge modules in nixpkgs. Having options for every config setting has some serious downsides, check out NixOS/rfcs#42 if you haven't seen it already.

basvandijk commented:
@Infinisil thanks for pointing me to that RFC. I didn't see it yet. You raise some interesting points. I guess I'll have to make my mind up what my opinion is on the matter.

basvandijk commented:
@GrahamcOfBorg test prometheus2

basvandijk commented:
@GrahamcOfBorg test prometheus2

basvandijk committed:
thanos: 0.4.0 -> 0.5.0-rc.0

basvandijk committed:
nixos/thanos: allow overriding arguments to the thanos subcommands

basvandijk commented:
@GrahamcOfBorg test prometheus2

basvandijk commented:
@Infinisil regarding https://github.com/NixOS/rfcs/pull/42, I understand the motivation for it: NixOS modules that encode every setting as a NixOS option are, to quote the RFC, "time consuming to implement, tedious to review and hard to maintain". This is true. I do think though that the time invested in implementing, reviewing and maintaining the module will benefit the user significantly. She'll get a module which is
* discoverable: every setting is a NixOS options which is searchable and documented
* safe to use: eval-time syntax- and type-checking for every option.

OTOH reducing the time spent implementing, reviewing and maintaining modules is also important so I'm willing to investigate a `config`-based approach. Regarding that I have a few questions:

* Thanos doesn't work with a configuration file but accepts command line arguments instead. It's unclear to me how to map an option of type `types.json` to a list of command line arguments.

  For most type of options the mapping is not that hard. For example we can have:
  `{ store.sd-interval = "10m"; store.unhealthy-timeout = "3m"; }` -> 
  `[ ''--store.sd-interval="10m"'' ''store.unhealthy-timeout="3m"'' ]`

  But what to do with repeatable options like:

      --store=<store> ...        Addresses of statically configured store API
                                 servers (repeatable). The scheme may be
                                 prefixed with 'dns+' or 'dnssrv+' to detect
                                 store API servers through respective DNS
                                 lookups.  

  This PR solves this by having the option: 
  `store.addresses = mkOption { type = types.listOf types.str; ... }` 
  This option will then get converted to a list of multiple `--store="<address>"` arguments.

  Of course we could implement the same kind of logic with the `config` setting where we would check if `config` has a `store.addresses` field of type list of string and then converting that. However, in contrast to the NixOS option, that field would not be discoverable.

* Similarly to the above list-based arguments Thanos has key-value-based arguments like:

      --label=<name>="<value>" ...
                                 Labels to be applied to all generated metrics
                                 (repeated). Similar to external labels for
                                 Prometheus, used to identify ruler and its
                                 blocks as unique source.

  How would we specify and convert these? Maybe by checking for a field named `label` in the `config` which should be an attribute set, the converting the name-value pairs to `--label=<name>="<value>"`. This could work but is also not very discoverable.

mmahut commented:
I've tested the binary and this works as expected.

kalbasit commented on nixos/modules/services/monitoring/thanos.nix:
```diff
@@ -0,0 +1,770 @@
+  listToArgs   = opt : vs  : map               (v: ''--${opt}="${v}"'') vs;
+  attrsToArgs  = opt : kvs : mapAttrsToList (k: v: ''--${opt}=${k}=\"${v}\"'') kvs;
+
+  mkParamDef = type : default : description : mkParam type (description + ''
```
I'm not really sure about the space before the column. Although this does look nice, I have not seen many examples of it.

basvandijk replied:
I see this is indeed not a common style in nixpkgs. I'll remove the space before all colons.

kalbasit commented on nixos/modules/services/monitoring/thanos.nix:
```diff
@@ -0,0 +1,770 @@
+
+    gcloudtrace = {
+
+      gcloudtrace.project = mkParam types.str ''
```
isn't redundant to have to do `gcloudtrace.gcloudtrace.project = ...`?

basvandijk replied:
The outer `gcloudtrace` is just so we can share these parameters with multiple services. A user will set these options using something like:
`services.thanos.query.gcloudtrace.project = ...` or
`services.thanos.store.gcloudtrace.project = ...`

kalbasit commented on nixos/modules/services/monitoring/thanos.nix:
```diff
@@ -0,0 +1,770 @@
+    (mkIf cfg.sidecar.enable {
+      systemd.services.thanos-sidecar = {
+        wantedBy = [ "multi-user.target" ];
+        after    = [ "network.target" "prometheus2.service" ];
```
should we enable prometheus (`services.prometheus.enable`) or write an assertion for it?

basvandijk replied:
I think an assertion is an excellent idea! I'll add it.

kalbasit replied:
Awesome! Thank you for adding the assertions. Not a blocker, but it would be great to add a test for the assertions.

kalbasit commented on pkgs/servers/monitoring/thanos/default.nix:
```diff
@@ -0,0 +1,46 @@
+    runCommandNoCC "bzr-no-cert-validation" {
+      inherit bazaar;
+      buildInputs = [ makeWrapper ];
+    } "makeWrapper $bazaar/bin/bzr $out/bin/bzr --add-flags -Ossl.cert_reqs=none";
```
is there a way this can go into go-modules instead?

basvandijk replied:
We could add it to `nativeBuildInputs` by default but that would cause an unnecessary build dependency for all go-modules (except for `thanos` of course).

However [`git` is also added by default](https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/go-modules/generic/default.nix#L39) so there's precedent. OTOH `git` is used by most go-modules while `bzr` isn't. 

In general I think it's worthwhile to have support for overriding go-modules.

kalbasit replied:
I understand the need to be able to support overriding the go-modules and I would leave that. I'm just thinking that any package using `bzr` will have to copy your hack over to their package and most likely spend time figuring out why they're getting a TLS error. We could maybe control it with a flag defaulting to false? Something like `enable-bzr ? false`?

basvandijk replied:
Apparently `thanos-0.5.0` doesn't depend on `bzr` anymore so I removed the `overrideModAttrs` setting. 

It would still be nice to add a `enableBzr` flag but this can be done in a separate PR.

kalbasit commented on pkgs/servers/monitoring/thanos/default.nix:
```diff
@@ -0,0 +1,46 @@
+
+  meta = with stdenv.lib; {
+    description = "Highly available Prometheus setup with long term storage capabilities";
+    homepage = https://github.com/improbable-eng/thanos;
```
```suggestion
    homepage = "https://github.com/improbable-eng/thanos";
```

https://github.com/NixOS/rfcs/pull/45

basvandijk replied:
Good point. I'll change it.

kalbasit reviewed (changes requested):
I have not tested yet, unsure if I can afford to do so myself. This is a pure CR.

basvandijk committed:
thanos: don't use URL literals (RFC-0045)

basvandijk committed:
nixos/thanos: assert that prometheus2 is running and has labels set

basvandijk committed:
nixos/thanos: code style: don't use a space before a colon

basvandijk committed:
thanos: 0.5.0-rc.0 -> 0.5.0

basvandijk committed:
thanos: remove bzr because 0.5.0 apparently doesn't need it anymore

I'll leave in the support for overrideModAttrs because that can be
useful for other go packages.

basvandijk commented:
@GrahamcOfBorg test prometheus2

basvandijk commented:
The test fails on `aarch64-linux` because of insufficient disk space:

```
store# [  126.911144] thanos[739]: level=error ts=2019-06-16T14:00:26.59870538Z caller=main.go:182 msg="running command failed" err="error executing compaction: first pass of downsampling failed: create dir: mkdir /var/lib/thanos-compact/downsample: no space left on device"
store# [  126.942655] systemd[1]: thanos-compact.service: Main process exited, code=exited, status=1/FAILURE
```

I'll increase the disk space on `store` and run the test again...

basvandijk committed:
nixos/tests/prometheus-2.nix: increase diskSize of the store machine

This is to fix the following error in the test on aarch64-linux:

store# [  126.911144] thanos[739]: level=error ts=2019-06-16T14:00:26.59870538Z caller=main.go:182 msg="running command failed" err="error executing compaction: first pass of downsampling failed: create dir: mkdir /var/lib/thanos-compact/downsample: no space left on device"
store# [  126.942655] systemd[1]: thanos-compact.service: Main process exited, code=exited, status=1/FAILURE

basvandijk commented:
@GrahamcOfBorg test prometheus2

basvandijk committed:
thanos: 0.5.0 -> 0.6.0

basvandijk commented:
@GrahamcOfBorg test prometheus2

basvandijk commented:
@GrahamcOfBorg test prometheus2
